### PR TITLE
WIP: Kubeadm reset should cleanup control-plane nodes properly

### LINKED ISSUE #1312: Kubeadm reset should cleanup control-plane nodes properly
Kubeadm reset should take care of removing control-plane nodes from the ClusterStatus status ConfigMap

### PR DESCRIPTION
**Pull request by Klaven**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

/kind bug

**What this PR does / why we need it**:
Fixes bug in https://github.com/kubernetes/kubeadm/issues/1312 failure to remove control plane node from Config.

**Which issue(s) this PR fixes**:

Fixes kubernetes/kubeadm#1312

**Special notes for your reviewer**:
I'm looking for input on this codebase, I know this code needs refactored, but I was trying to do the minimum needed to fix the bug, I thought I would write up a issue about refactoring the code surrounding this? Anyhow, I want  your all thoughts before I push forward for a final PR. I can do more refactors and make the code a bit cleaner, but then the PR is not just fixing the bug, it's also a refactor.

Reasons in WIP:
- [ ] Looking for input from others
- [ ] Need to test in a real cluster (wanted to get opinions)
- [ ] Squash (will do just before removing WIP)

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by Klaven** [comment]
@fabriziopandini @luxas @ereslibre If you all don't mind, This is the start to a fix. Would love your alls thoughts on it! I struggled with this because part of me wanted to do a large refactor. but that would have been a lot of code change that was not fixing the bug. So I tried to fix it with as few changes to the code base and I thought I would write up a refactoring bug later today. I also need help with how I should test this. 

Thank you.

File: cmd/kubeadm/app/cmd/reset.go [hunk 1]
```diff
@@ -131,6 +133,20 @@ func (r *Reset) Run(out io.Writer, client clientset.Interface) error {
...
+	// Check if control plane node
+	klog.V(1).Infof("[reset] checking if control plane node")
+	if isControlPlane(cfg) {
+		klog.V(1).Infof("[reset] removing control plane node from cluster")
+		uploadconfig.UploadRemoveConfiguration(cfg, client)
+	}
+
+	// Try to stop etcd
 	klog.V(1).Infof("[reset] checking for etcd config")
 	etcdDataDir, err := getEtcdDataDir(etcdManifestPath, client)
```
Comment (Klaven) [inline cmd/kubeadm/app/cmd/reset.go]: The only thing this needs the `client` for is to get the config..... what I did above? seems like a waste for it to do it again... but I did not have to change this code so I left it.

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go [hunk 2]
```diff
@@ -37,6 +37,53 @@ const (
 	NodesKubeadmConfigClusterRoleName = "kubeadm:nodes-kubeadm-config"
 )
 
+// UploadRemoveConfiguration uploads a new configuration with the clinet removed
+func UploadRemoveConfiguration(cfg *kubeadmapi.InitConfiguration, client clientset.Interface) error {
```
Comment (Klaven) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: All of this needs refactoring I think. but this was the cleanest way to do this without really touching "working code" I'm very much open to ideas here.

File: cmd/kubeadm/app/cmd/reset.go [hunk 3]
```diff
@@ -295,3 +311,19 @@ func resetConfigDir(configPathDir, pkiPathDir string) {
 		}
 	}
 }
+
+//Checks if is control plane
```
Comment (nanoscopic) [inline cmd/kubeadm/app/cmd/reset.go]: Missing space before 'Checks'. Also this comment is not adding any additional information beyond the name of the function. It is no different from just leaving the comment out. Should make it more descriptive english or drop this comment.

File: cmd/kubeadm/app/cmd/reset.go [hunk 4]
```diff
@@ -295,3 +311,19 @@ func resetConfigDir(configPathDir, pkiPathDir string) {
...
+func isControlPlane(cfg *kubeadmapi.InitConfiguration) bool {
+
```
Comment (nanoscopic) [inline cmd/kubeadm/app/cmd/reset.go]: Extra blank lines at start and end of function; should be dropped to match the style of the rest of the file.

File: cmd/kubeadm/app/cmd/reset.go [hunk 5]
```diff
@@ -295,3 +311,19 @@ func resetConfigDir(configPathDir, pkiPathDir string) {
...
+	if cfg == nil {
+		return false
+	}
+
+	filepath := kubeadmconstants.GetStaticPodFilepath(kubeadmconstants.KubeAPIServer, kubeadmconstants.GetStaticPodDirectory())
```
Comment (nanoscopic) [inline cmd/kubeadm/app/cmd/reset.go]: Comment here stating that only the control plane has a static pod directory and that is how it is being determined if this is the control plane could be nice.

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go [hunk 6]
```diff
@@ -37,6 +37,53 @@ const (
 	NodesKubeadmConfigClusterRoleName = "kubeadm:nodes-kubeadm-config"
 )
 
+// UploadRemoveConfiguration uploads a new configuration with the clinet removed
```
Comment (nanoscopic) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: client is misspelled here

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go (hunk 2 above)
Comment (nanoscopic) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: Comment of function should say that nil is returned on success and the error if any on error.

**Review by nanoscopic — COMMENTED** [review]
Minor styling comments and ideas for additional code comments.

**Review by neolit123 — COMMENTED** [review]
@Klaven please rename the PR and commit titles to not include the related issue, but instead explain what the PR is changing.

**Comment by neolit123** [comment]
the `Fixes` line should look like this:
`Fixes kubernetes/kubeadm#1312`

**Comment by Klaven** [comment]
> @Klaven please rename the PR and commit titles to not include the related issue, but instead explain what the PR is changing.

I will when it's a real PR. right now I was just hopping to get some ideas about the code and the fix. do you have any thoughts on it @neolit123 . Also would you know of a good way to provide good test around this? thanks!

File: cmd/kubeadm/app/cmd/reset.go [hunk 7]
```diff
@@ -132,6 +134,20 @@ func (r *Reset) Run(out io.Writer, client clientset.Interface) error {
...
 	etcdManifestPath := filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.ManifestsSubDirName, "etcd.yaml")
 
+	cfg, err := configutil.FetchConfigFromFileOrCluster(client, os.Stdout, "reset", "", false)
+
+	if err != nil && cfg != nil {
+		cfg = nil
+	}
+
+	// Check if control plane node
+	klog.V(1).Infof("[reset] checking if control plane node")
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/cmd/reset.go]: Writing this log entry IMO doesn't add value...

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go [hunk 8]
```diff
@@ -37,6 +37,53 @@ const (
 	NodesKubeadmConfigClusterRoleName = "kubeadm:nodes-kubeadm-config"
 )
 
+// UploadRemoveConfiguration uploads a new configuration with the client removed
+func UploadRemoveConfiguration(cfg *kubeadmapi.InitConfiguration, client clientset.Interface) error {
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: What about calling this func `ResetClusterStatusForNode` instead of `UploadRemoveConfiguration`?
if possible, invert the order of parameters (first `client`, then `cfg`); additionally I the next comments apply, don't think we need all the cfg, but only the node name.

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go [hunk 9]
```diff
@@ -37,6 +37,53 @@ const (
...
+	fmt.Printf("[reset] removing the configuration stored in the ConfigMap %q in the %q Namespace\n", kubeadmconstants.KubeadmConfigConfigMap, metav1.NamespaceSystem)
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: What about "removing node info from the kubeadm-config config map"?

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go (hunk 8 above)
Comment (fabriziopandini) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: I agree with you this function should be refactored. What about:

- Read the ConfigMap (prior art [here](https://github.com/kubernetes/kubernetes/blob/c6977e5aac7b2c3645825b4b21512d4b93704cfa/cmd/kubeadm/app/util/config/cluster.go#L236))
- Unmarshal the ClusterStatus bit (reuse [unmarshalClusterStatus](https://github.com/kubernetes/kubernetes/blob/c6977e5aac7b2c3645825b4b21512d4b93704cfa/cmd/kubeadm/app/util/config/cluster.go#L252) func ; requires make it public)
- Delete the node entry from ClusterStatus
- Marshal back the ClusterStatus and Save

The main advantage of this approach is that we are not touching at all the ClusterConfiguration part.
Wdyt

**Review by fabriziopandini — COMMENTED** [review]
@Klaven thanks for sharing this WIP in advance!
Below my comments

**Comment by fabriziopandini** [comment]
@Klaven when do you plan to get back working on this PR? (I'm updating state for Tracking issue for Certificates copy for join --control-plane)

File: cmd/kubeadm/app/cmd/reset.go [hunk 10]
```diff
@@ -132,6 +134,20 @@ func (r *Reset) Run(out io.Writer, client clientset.Interface) error {
 	// Only clear etcd data when using local etcd.
 	etcdManifestPath := filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.ManifestsSubDirName, "etcd.yaml")
 
+	cfg, err := configutil.FetchConfigFromFileOrCluster(client, os.Stdout, "reset", "", false)
```
Comment (rosti) [inline cmd/kubeadm/app/cmd/reset.go]: Please, wait for #74410 to land in and leverage on work from that.

File: cmd/kubeadm/app/cmd/reset.go [hunk 11]
```diff
@@ -295,3 +311,17 @@ func resetConfigDir(configPathDir, pkiPathDir string) {
 		}
 	}
 }
+
+// isControlPlane Checks if is control plane by checking if they have the cfg and if the KubeAPIServer static pod exists
+func isControlPlane(cfg *kubeadmapi.InitConfiguration) bool {
+	if cfg == nil {
```
Comment (rosti) [inline cmd/kubeadm/app/cmd/reset.go]: This may be true on a broken control plane node. Hence I think, that the real test is the existence of the API Server pod spec and this should be removed.

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go (hunk 8 above)
Comment (rosti) [reply cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: +1 on that.

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go (hunk 9 above)
Comment (rosti) [reply cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: +1

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go (hunk 8 above)
Comment (rosti) [reply cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: Additionally, we should take only a node name string (instead of an InitConfiguration) as a func parameter to achieve what @fabriziopandini suggested.

File: cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go [hunk 12]
```diff
@@ -65,7 +112,6 @@ func UploadConfiguration(cfg *kubeadmapi.InitConfiguration, client clientset.Int
 	if clusterStatus.APIEndpoints == nil {
 		clusterStatus.APIEndpoints = map[string]kubeadmapi.APIEndpoint{}
 	}
-	clusterStatus.APIEndpoints[cfg.NodeRegistration.Name] = cfg.LocalAPIEndpoint
```
Comment (rosti) [inline cmd/kubeadm/app/phases/uploadconfig/uploadconfig.go]: I may be missing something, but why is this removed?

**Review by rosti — CHANGES_REQUESTED** [review]
Thanks @Klaven !
Do you have some time to work on this one before code freeze (Mar 7th)?
If not I can pick this up.

**Comment by pytimer** [comment]
@Klaven @rosti I can take it if you don't have time t deal with this PR.

**Comment by rosti** [comment]
@pytimer I just got in touch with @Klaven, he is going to finish this.

**Comment by pytimer** [comment]
Okay, i know. Thanks, @rosti .

**Comment by neolit123** [comment]
sending a new PR later today.
will address the comments here.

/close
